modAionImpl/src/org/aion/zero/impl/sync/msg/ResponseTrieData.java

```diff
@@ -0,0 +1,259 @@
+        this.nodeValue = nodeValue;
+        this.referencedNodes = referencedNodes;
+        this.dbType = dbType;
+    }
+
+    /**
+     * Decodes a message into a trie node response.
+     *
+     * @param message a {@code byte} array representing a response to a trie node request.
+     * @return the decoded trie node response
+     * @implNote The decoder must return {@code null} if any of the component values are {@code
+     *     null} or invalid.
+     */
+    public static ResponseTrieData decode(final byte[] message) {
+        if (message == null || message.length == 0) {
+            return null;
+        } else {
+            RLPList list = (RLPList) RLP.decode2(message).get(0);
+            if (list.size() != 4) {
+                return null;
+            } else {
+                // decode the key
+                byte[] hash = list.get(0).getRLPData();
+                if (hash.length != 32) {
+                    return null;
+                }
+
+                // decode the value
+                byte[] value = list.get(1).getRLPData();
+                if (value.length == 0) {
+                    return null;
+                }
+
+                // decode the referenced nodes
+                RLPElement referenced = list.get(2);
+                if (!(referenced instanceof RLPList)) {
+                    return null;
+                }
+                Map<ByteArrayWrapper, byte[]> nodes = decodeReferencedNodes((RLPList) referenced);
+                if (nodes == null) {
+                    return null;
+                }
+
+                // decode the db type
+                byte[] type = list.get(3).getRLPData();
+                DatabaseType dbType;
+                try {
+                    dbType = DatabaseType.valueOf(new String(type));
+                } catch (IllegalArgumentException e) {
+                    return null;
+                }
+
+                return new ResponseTrieData(ByteArrayWrapper.wrap(hash), value, nodes, dbType);
+            }
+        }
+    }
+
+    /**
+     * Decodes the list of key-value pair encodings into a map representation.
+     *
+     * @return a map of all the key-value pair encodings
+     */
+    private static Map<ByteArrayWrapper, byte[]> decodeReferencedNodes(RLPList referenced) {
+        if (referenced.isEmpty()) {
+            return Collections.emptyMap();
+        }
+
+        RLPList current;
+        byte[] hash, value;
+        Map<ByteArrayWrapper, byte[]> nodes = new HashMap<>();
+
+        for (RLPElement pair : referenced) {
+            if (!(pair instanceof RLPList)) {
+                return null;
+            }
+            current = (RLPList) pair;
+
+            // decode the key
+            hash = current.get(0).getRLPData();
+            if (hash.length != 32) {
+                return null;
+            }
+
+            // decode the value
+            value = current.get(1).getRLPData();
+            if (value.length == 0) {
```

Comment:
could also do ! value.isEmpty()

Reply:
The `value` is a `byte[]`. Don't have an `isEmpty()` method in this case.